components/core/src/package/target.rs

```diff
@@ -0,0 +1,197 @@
+    ///
+    /// Errors:
+    /// * InvalidPackageTargetPlatform
+    /// * InvalidPackageTargetArchitecture
+    pub fn new<T: Into<String>>(platform: T, architecture: T) -> Self {
+        let valid_platform = PackageTarget::supported_platform(platform).unwrap();
+        let valid_arch = PackageTarget::supported_architecture(architecture).unwrap();
+        PackageTarget {
+            platform: valid_platform,
+            architecture: valid_arch,
+        }
+    }
+
+    fn supported_platform<T: Into<String>>(value: T) -> Result<String> {
+        let platform = value.into();
+        let valid_platforms = vec!["linux", "windows",];
+        if valid_platforms.iter().any(|&x| x == platform) {
+            Ok(platform)
+        } else {
+            Err(Error::InvalidPackageTargetPlatform(platform))
+        }
+    }
+
+    fn supported_architecture<T: Into<String>>(value: T) -> Result<String> {
+        let architecture = value.into();
+        let valid_architectures = vec!["x86_64"];
+        if valid_architectures.iter().any(|&x| x == architecture) {
+            Ok(architecture)
+        } else {
+            Err(Error::InvalidPackageTargetArchitecture(architecture))
+        }
+    }
+}
+
+impl Target for PackageTarget {
+    fn platform(&self) -> &str {
+        &self.platform
+    }
+    fn architecture(&self) -> &str {
+        &self.architecture
+    }
+    fn valid_for(&self, uname: Uname) -> Result<()> {
+        let architecture = uname.machine.trim().to_lowercase();
+        let platform = uname.sys_name.trim().to_lowercase();
+        if self.architecture == architecture && self.platform == platform {
+            Ok(())
+        } else {
+            Err(Error::TargetMatchError(format!("Package target ({}-{}) does not match system \
+                                                 target ({}-{}).",
+                                                self.architecture,
+                                                self.platform,
+                                                architecture,
+                                                platform,)))
+        }
+    }
+}
+
+impl Default for PackageTarget {
+    fn default() -> PackageTarget {
+        let current_system = uname().unwrap();
+        PackageTarget::new(current_system.sys_name.trim().to_lowercase(),
+                           current_system.machine.trim().to_lowercase())
+    }
+}
+
+impl fmt::Display for PackageTarget {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}-{}", self.architecture, self.platform)
+    }
+}
+
+impl AsRef<PackageTarget> for PackageTarget {
+    fn as_ref(&self) -> &PackageTarget {
+        self
+    }
+}
+
+impl FromStr for PackageTarget {
+    type Err = Error;
+
+    fn from_str(value: &str) -> result::Result<Self, Self::Err> {
+        let items: Vec<&str> = value.split("-").collect();
+        let (architecture, platform) = match items.len() {
+            2 => (items[0], items[1]),
```

Comment:
In here if you go the route of using enums to represent the Architecture and Platform and have implemented `FromStr` on them, you can use that to chain an error upwards on error in this FromStr implementation